File: apps/mobile/modules/task-handoff-browser/ios/MobileBrowserContextManager.swift
import Foundation
import CryptoKit
import Network
import OSLog
import WebKit

private let browserContextLogger = Logger(subsystem: "com.taskhandoff.mobile.dev", category: "task-handoff-browser")

internal actor MobileBrowserContextManager {
  static let shared = MobileBrowserContextManager()

  private struct Context {
    let id: String
    let key: String
    let dataStore: WKWebsiteDataStore
    let channel: BrowserTunnelChannel
    let socks: BrowserSocksServer
    var references: Int
  }

  private var contextsById: [String: Context] = [:]
  private var contextIdByKey: [String: String] = [:]
  private var preparedContexts = 0
  private var releasedContexts = 0

  @available(iOS 17.0, *)
  func prepare(_ input: PrepareBrowserContextRecord) async throws -> String {
    let key = "\(input.controlPlaneId)\u{0}\(input.instanceId)"
    if let id = contextIdByKey[key], var context = contextsById[id] {
      context.references += 1
      contextsById[id] = context
      browserDiagnostic("context reused id=\(id) references=\(context.references)")
      return id
    }

    // The token remains in this stack frame only. The tunnel consumes it while
    // establishing the relay and never includes it in the opaque context id.
    guard !input.token.isEmpty, let relayURL = URL(string: input.relayUrl), ["ws", "wss"].contains(relayURL.scheme?.lowercased()) else {
      throw BrowserContextUnavailableException()
    }
    let channel = BrowserTunnelChannel(url: relayURL, token: input.token)
    try await channel.connect()
    let socks = BrowserSocksServer(channel: channel)
    let address: BrowserSocksServer.Address
    do {
      address = try await socks.start()
    } catch {
      await channel.close()
      throw error
    }
    let dataStore = await MainActor.run {
      let store = WKWebsiteDataStore(forIdentifier: profileIdentifier(key))
      var proxy = ProxyConfiguration(socksv5Proxy: .hostPort(host: address.host, port: address.port))
      proxy.allowFailover = false
      store.proxyConfigurations = [proxy]
      browserDiagnostic("iOS WebView SOCKS proxy configured host=\(address.host) port=\(address.port) profile=\(profileIdentifier(key))")
      return store
    }
    let id = UUID().uuidString.lowercased()
    let context = Context(id: id, key: key, dataStore: dataStore, channel: channel, socks: socks, references: 1)
    contextsById[id] = context
    contextIdByKey[key] = id
    preparedContexts += 1
    browserDiagnostic("context prepared id=\(id) instance=\(input.instanceId)")
    return id
  }

  func dataStore(contextId: String) -> WKWebsiteDataStore? {
    contextsById[contextId]?.dataStore
  }

  func activate(contextId: String) throws {
    guard contextsById[contextId] != nil else { throw BrowserContextUnavailableException() }
  }

  func release(contextId: String) async {
    guard var context = contextsById[contextId] else { return }
    context.references -= 1
    if context.references > 0 {
      contextsById[contextId] = context
      return
    }
    contextsById.removeValue(forKey: contextId)
    contextIdByKey.removeValue(forKey: context.key)
    releasedContexts += 1
    browserDiagnostic("context released id=\(contextId)")
    await context.socks.close()
    await context.channel.close()
  }

  func releaseAll() async {
    let active = Array(contextsById.values)
    contextsById.removeAll()
    contextIdByKey.removeAll()
    releasedContexts += active.count
    for context in active {
      await context.socks.close()
      await context.channel.close()
    }
  }

  func diagnostics() -> [String: Int] {
    ["activeContexts": contextsById.count, "preparedContexts": preparedContexts, "releasedContexts": releasedContexts]
  }
}

private func browserDiagnostic(_ message: String) {
  NSLog("[task-handoff-browser] %@", message)
  browserContextLogger.info("\(message, privacy: .public)")
}

private func profileIdentifier(_ key: String) -> UUID {
  let digest = Array(SHA256.hash(data: Data(key.utf8)).prefix(16))
  let value = digest.map { String(format: "%02x", $0) }.joined()
  return UUID(uuidString: "\(value.prefix(8))-\(value.dropFirst(8).prefix(4))-\(value.dropFirst(12).prefix(4))-\(value.dropFirst(16).prefix(4))-\(value.dropFirst(20).prefix(12))")!
}

File: apps/mobile/modules/task-handoff-browser/android/src/main/java/expo/modules/taskhandoffbrowser/MobileBrowserContextManager.kt
package expo.modules.taskhandoffbrowser

import androidx.webkit.ProfileStore
import androidx.webkit.ProxyConfig
import androidx.webkit.ProxyController
import androidx.webkit.WebViewFeature
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.net.URI
import java.security.MessageDigest
import java.util.UUID
import kotlin.coroutines.resume

internal object MobileBrowserContextManager {
  private data class Context(
    val id: String,
    val key: String,
    val profileName: String,
    val socksPort: Int,
    val channel: BrowserTunnelChannel,
    val socks: BrowserSocksServer,
    var references: Int,
  )

  private val mutex = Mutex()
  private val contextsById = mutableMapOf<String, Context>()
  private val contextIdByKey = mutableMapOf<String, String>()
  private var preparedContexts = 0
  private var releasedContexts = 0

  suspend fun prepare(input: PrepareBrowserContextRecord): String = mutex.withLock {
    requireCapabilities()
    val relay = runCatching { URI(input.relayUrl) }.getOrNull()
    if (input.controlPlaneId.isBlank() || input.instanceId.isBlank() || input.token.isBlank() || relay?.scheme !in setOf("ws", "wss")) {
      throw BrowserContextUnavailableException()
    }
    val key = "${input.controlPlaneId}\u0000${input.instanceId}"
    contextIdByKey[key]?.let { id ->
      contextsById[id]?.let { context ->
        context.references += 1
        return id
      }
    }
    if (contextsById.isNotEmpty()) {
      throw BrowserPlatformUnsupportedException("Close the active instance Browser context before switching instances.")
    }

    val channel = BrowserTunnelChannel(input.relayUrl, input.token)
    var socks: BrowserSocksServer? = null
    try {
      channel.connect()
      socks = BrowserSocksServer(channel)
      val port = socks.start()
      val profileName = deriveProfileName(key)
      withContext(Dispatchers.Main) { ProfileStore.getInstance().getOrCreateProfile(profileName) }
      setProxy(port)
      val id = UUID.randomUUID().toString()
      val context = Context(id, key, profileName, port, channel, socks, 1)
      contextsById[id] = context
      contextIdByKey[key] = id
      preparedContexts += 1
      id
    } catch (error: Throwable) {
      socks?.closeAndJoin()
      channel.closeAndJoin()
      throw error
    }
  }

  suspend fun profileName(contextId: String): String? = mutex.withLock { contextsById[contextId]?.profileName }

  data class DownloadContext(val profileName: String, val socksPort: Int)
  suspend fun downloadContext(contextId: String): DownloadContext? = mutex.withLock {
    contextsById[contextId]?.let { DownloadContext(it.profileName, it.socksPort) }
  }

  suspend fun release(contextId: String) = mutex.withLock {
    val context = contextsById[contextId] ?: return
    context.references -= 1
    if (context.references > 0) return
    contextsById.remove(contextId)
    contextIdByKey.remove(context.key)
    releasedContexts += 1
    clearProxy()
    context.socks.closeAndJoin()
    context.channel.closeAndJoin()
  }

  suspend fun releaseAll() = mutex.withLock {
    val active = contextsById.values.toList()
    contextsById.clear()
    contextIdByKey.clear()
    releasedContexts += active.size
    if (active.isNotEmpty()) clearProxy()
    for (context in active) {
      context.socks.closeAndJoin()
      context.channel.closeAndJoin()
    }
  }

  suspend fun diagnostics(): Map<String, Any?> = mutex.withLock { mapOf(
    "activeContexts" to contextsById.size,
    "preparedContexts" to preparedContexts,
    "releasedContexts" to releasedContexts,
  ) }

  private fun requireCapabilities() {
    if (!WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE) ||
      !WebViewFeature.isFeatureSupported(WebViewFeature.MULTI_PROFILE)) {
      throw BrowserPlatformUnsupportedException("Android WebView proxy override and isolated profile support are required.")
    }
  }

  private suspend fun setProxy(port: Int) = suspendCancellableCoroutine { continuation ->
    val config = ProxyConfig.Builder().addProxyRule("socks://127.0.0.1:$port").build()
    ProxyController.getInstance().setProxyOverride(config, Runnable::run) { if (continuation.isActive) continuation.resume(Unit) }
  }

  private suspend fun clearProxy() = suspendCancellableCoroutine { continuation ->
    ProxyController.getInstance().clearProxyOverride(Runnable::run) { if (continuation.isActive) continuation.resume(Unit) }
  }

  private fun deriveProfileName(key: String): String {
    val digest = MessageDigest.getInstance("SHA-256").digest(key.toByteArray())
    return "taskhandoff-${digest.take(16).joinToString("") { "%02x".format(it) }}"
  }
}
